From: William Horvath <william@horvath.blog>
Date: Fri, 05 Jul 2024 08:10:03 +0000
Subject: [PATCH] alsa,pulse,mmdevapi: Support custom period and buffer sizes, and respect PulseAudio settings.

Relevant environment variables are: 
ALSA_EXTRA_PAD (units: usecs*10; default is 40000, can be set to 0, adds extra padding)
STAGING_AUDIO_PERIOD (units: usecs*10)
STAGING_AUDIO_DURATION (units: usecs*10)

Also, silence pcm.c underrun errors.

diff --git a/dlls/winealsa.drv/alsa.c b/dlls/winealsa.drv/alsa.c
index 11111111111..11111111111 100644
--- a/dlls/winealsa.drv/alsa.c
+++ b/dlls/winealsa.drv/alsa.c
@@ -82,10 +82,10 @@ struct alsa_stream
     pthread_mutex_t lock;
 };
 
-#define EXTRA_SAFE_RT 40000
+REFERENCE_TIME EXTRA_SAFE_RT = 40000;
 
-static const REFERENCE_TIME def_period = 100000;
-static const REFERENCE_TIME min_period = 50000;
+REFERENCE_TIME def_period = 1000;
+REFERENCE_TIME min_period = 500;
 
 static const WCHAR drv_keyW[] = {'S','o','f','t','w','a','r','e','\\',
     'W','i','n','e','\\','D','r','i','v','e','r','s','\\',
@@ -810,6 +810,19 @@ static NTSTATUS alsa_create_stream(void *args)
 
     params->result = S_OK;
 
+    const char* esrtenv = getenv("ALSA_EXTRA_PAD");
+
+    REFERENCE_TIME esrtval;
+
+    if (esrtenv) {
+        esrtval = atol(esrtenv);
+        if (esrtval >= 0) EXTRA_SAFE_RT = esrtval; // failsafe
+        MESSAGE("ALSA padding:\t%llu\n", EXTRA_SAFE_RT);
+    }
+    else {
+        TRACE("ALSA padding:\t%llu\t(default)\n", EXTRA_SAFE_RT);
+    }
+
     stream = calloc(1, sizeof(*stream));
     if(!stream){
         params->result = E_OUTOFMEMORY;
@@ -1008,10 +1021,10 @@ static NTSTATUS alsa_create_stream(void *args)
 
     pthread_mutex_init(&stream->lock, NULL);
 
-    TRACE("ALSA period: %lu frames\n", stream->alsa_period_frames);
-    TRACE("ALSA buffer: %lu frames\n", stream->alsa_bufsize_frames);
-    TRACE("MMDevice period: %u frames\n", stream->mmdev_period_frames);
-    TRACE("MMDevice buffer: %u frames\n", stream->bufsize_frames);
+    MESSAGE("PIPEWIRE_ALSA period: %lu frames\n", stream->alsa_period_frames);
+    MESSAGE("PIPEWIRE_ALSA buffer: %lu frames\n", stream->alsa_bufsize_frames);
+    MESSAGE("MMDevAPI period: %u frames\n", stream->mmdev_period_frames);
+    MESSAGE("MMDevAPI buffer: %u frames\n", stream->bufsize_frames);
 
 exit:
     free(sw_params);
@@ -1280,7 +1295,7 @@ static snd_pcm_sframes_t alsa_write_best_effort(struct alsa_stream *stream, BYTE
         WARN("writei failed, recovering: %ld (%s)\n", written,
                 snd_strerror(written));
 
-        ret = snd_pcm_recover(stream->pcm_handle, written, 0);
+        ret = snd_pcm_recover(stream->pcm_handle, written, 1);
         if(ret < 0){
             WARN("Could not recover: %d (%s)\n", ret, snd_strerror(ret));
             return ret;
@@ -1454,7 +1469,7 @@ static void alsa_read_data(struct alsa_stream *stream)
 
         WARN("read failed, recovering: %ld (%s)\n", nread, snd_strerror(nread));
 
-        ret = snd_pcm_recover(stream->pcm_handle, nread, 0);
+        ret = snd_pcm_recover(stream->pcm_handle, nread, 1);
         if(ret < 0){
             WARN("Recover failed: %d (%s)\n", ret, snd_strerror(ret));
             return;
diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 16c9daaf59c..fdd9395e70b 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -943,8 +943,18 @@ static void pulse_probe_settings(pa_mainloop *ml, pa_context *ctx, int render, c
     if (stream)
         pa_stream_unref(stream);
 
-    if (length)
-        *def_period = *min_period = pa_bytes_to_usec(10 * length, &ss);
+    if (length) {
+        *min_period = pa_bytes_to_usec(10 * length, &ss);
+        // magic numbers...? +10 to make sure we round up instead of down
+        *def_period = (*min_period + (*min_period / 2)) + 10;
+    }
+
+    const char* penv = getenv("STAGING_AUDIO_PERIOD");
+    int val;
+    if (penv && (val = atoi(penv)) >= 0) {
+        *def_period = *min_period = val;
+        TRACE("Pulse audio period set to %d.\n", val);
+    }
 
     wfx->wFormatTag = WAVE_FORMAT_EXTENSIBLE;
     wfx->cbSize = sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX);
@@ -1246,7 +1256,7 @@ static HRESULT pulse_stream_connect(struct pulse_stream *stream, const char *pul
 
     /* PulseAudio will fill in correct values */
     attr.minreq = attr.fragsize = period_bytes;
-    attr.tlength = period_bytes * 3;
+    attr.tlength = period_bytes;
     attr.maxlength = stream->bufsize_frames * pa_frame_size(&stream->ss);
     attr.prebuf = pa_frame_size(&stream->ss);
     dump_attr(&attr);
@@ -1343,6 +1353,20 @@ static NTSTATUS pulse_create_stream(void *args)
     if (FAILED(hr))
         goto exit;
 
+    const char* penv = getenv("STAGING_AUDIO_PERIOD");
+    int periodval;
+    if ((stream->dataflow == eRender) && penv && (periodval = atoi(penv)) >= 0) {
+        params->period = periodval;
+        TRACE("Pulse audio period set to %d.\n", periodval);
+    }
+
+    const char* denv = getenv("STAGING_AUDIO_DURATION");
+    int durationval;
+    if ((stream->dataflow == eRender) && denv && (durationval = atoi(denv)) >= 0) {
+        params->duration = durationval;
+        TRACE("Pulse audio duration set to %d.\n", durationval);
+    }
+
     stream->def_period = params->period;
 
     stream->period_bytes = pa_frame_size(&stream->ss) * muldiv(params->period,
diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 77b1dcf6c4c..6837b0441b9 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -944,9 +944,10 @@ static void pulse_probe_settings(pa_mainloop *ml, pa_context *ctx, int render, c
         pa_stream_unref(stream);
 
     if (length) {
-        *min_period = pa_bytes_to_usec(10 * length, &ss);
-        // magic numbers...? +10 to make sure we round up instead of down
-        *def_period = (*min_period + (*min_period / 2)) + 10;
+        *min_period = pa_bytes_to_usec(10 * length, &ss);
+
+        // don't even ask
+        *def_period = pa_bytes_to_usec(10 * (length + (96 - (length % 96))), &ss);
     }
 
     const char* penv = getenv("STAGING_AUDIO_PERIOD");
diff --git a/dlls/mmdevapi/client.c b/dlls/mmdevapi/client.c
index 860c31f4b49..bb5228eb561 100644
--- a/dlls/mmdevapi/client.c
+++ b/dlls/mmdevapi/client.c
@@ -52,6 +52,8 @@ extern struct audio_session_wrapper *session_wrapper_create(struct audio_client
 
 static HANDLE main_loop_thread;
 
+static BOOL custom;
+
 void main_loop_stop(void)
 {
     if (main_loop_thread) {
@@ -105,7 +107,6 @@ static inline struct audio_client *impl_from_IAudioStreamVolume(IAudioStreamVolu
 static HRESULT get_periods(struct audio_client *client,
                            REFERENCE_TIME *def_period, REFERENCE_TIME *min_period)
 {
-    static const REFERENCE_TIME min_def_period = 100000; /* 10 ms */
     struct get_device_period_params params;
 
     params.device     = client->device_name;
@@ -115,8 +116,6 @@ static HRESULT get_periods(struct audio_client *client,
 
     wine_unix_call(get_device_period, &params);
 
-    if (def_period) *def_period = max(*def_period, min_def_period);
-
     return params.result;
 }
 
@@ -136,12 +135,10 @@ static HRESULT adjust_timing(struct audio_client *client, const BOOLEAN force_de
     TRACE("Device periods: %lu default and %lu minimum\n", (ULONG)def_period, (ULONG)min_period);
 
     if (mode == AUDCLNT_SHAREMODE_SHARED) {
-        if (*period == 0 || force_def_period)
+        if (!custom) {
             *period = def_period;
-        else if (*period < min_period)
-            return AUDCLNT_E_INVALID_DEVICE_PERIOD;
-        if (*duration < 3 * *period)
             *duration = 3 * *period;
+        }
     } else {
         const WAVEFORMATEXTENSIBLE *fmtex = (WAVEFORMATEXTENSIBLE *)fmt;
         if (fmtex->Format.wFormatTag == WAVE_FORMAT_EXTENSIBLE &&
@@ -166,7 +163,11 @@ static HRESULT adjust_timing(struct audio_client *client, const BOOLEAN force_de
         }
     }
 
-    TRACE("Adjusted duration %lu and period %lu\n", (ULONG)*duration, (ULONG)*period);
+    if (custom) {
+        MESSAGE("Set period %lu and duration %lu\n", (ULONG)*period, (ULONG)*duration);
+    } else {
+        TRACE("Adjusted duration %lu and period %lu\n", (ULONG)*duration, (ULONG)*period);
+    }
 
     return hr;
 }
@@ -378,6 +379,21 @@ static HRESULT stream_init(struct audio_client *client, const BOOLEAN force_def_
 
     if (mode != AUDCLNT_SHAREMODE_SHARED && mode != AUDCLNT_SHAREMODE_EXCLUSIVE)
         return E_INVALIDARG;
+    
+    if (mode == AUDCLNT_SHAREMODE_SHARED) {
+        const char* penv = getenv("STAGING_AUDIO_PERIOD");
+        const char* denv = getenv("STAGING_AUDIO_DURATION");
+        int periodval, durationval;
+        if (  (penv && denv) &&
+            (((periodval   = atoi(penv)) >= 0) &&
+             ((durationval = atoi(denv)) >= 0))
+           )
+        {
+            custom = TRUE;
+            period = periodval;
+            duration = durationval;
+        }
+    }
 
     if (flags & ~(AUDCLNT_STREAMFLAGS_CROSSPROCESS |
                   AUDCLNT_STREAMFLAGS_LOOPBACK |
@@ -1091,7 +1107,14 @@ static HRESULT WINAPI client_InitializeSharedAudioStream(IAudioClient3 *iface, D
     if (!format)
         return E_POINTER;
 
-    period = period_frames * (REFERENCE_TIME)10000000 / format->nSamplesPerSec;
+    const char* penv = getenv("STAGING_AUDIO_PERIOD");
+    int periodval;
+    if (penv && (periodval = atoi(penv)) >= 0) { // Custom
+        period = periodval;
+        TRACE("Audio period set to %d.\n", periodval);
+    } else { // Default
+        period = period_frames * (REFERENCE_TIME)10000000 / format->nSamplesPerSec;
+    }
 
     return stream_init(This, FALSE, AUDCLNT_SHAREMODE_SHARED, flags, 0, period, format, session_guid);
 }
